### DISCUSSION TITLE
Extracting and matching Python Error in Rust

### DISCUSSION THREAD

**Discussion by MohamedAbdeen21** [PyO3/pyo3#3759] [Questions] [answered]
[I'm relatively new to Rust, keep that in mind]

I want to extract the Python Error class (KeyError, TypeError,...) from PyErr. 

This is what I'm using at the moment
```rust
e.to_string().split_once(":").unwrap().0
```
It works, but feels a little hacky and wanted to see if there's a better approach.

When I debug the entire PyErr, I can see that it already has a field for it
```
pyo3_runtime.PanicException: PyErr { type: <class 'KeyError'>, value: KeyError('value2'), traceback: Some(<traceback object at 0x7f6518d38980>) }
```

I'm aware of the `is_instance_of` method, but is there a way without passing the Py token?

**Comment by adamreichold** [comment, marked as answer]
> I'm aware of the is_instance_of method, but is there a way without passing the Py token?

You definitely need a GIL token, but it should be easy to come by. Either you need to wrap your call to `is_instance_of` using `with_gil`, e.g.

```rust
fn is_key_error(err: &PyErr) -> bool {
    Python::with_gil(|py| {
        err.is_instance_of::<PyKeyError>(py)
    })
}
```

or better yet, you just add `py: Python` as an argument to your `#[pyfunction]` or `#[pymethod]` and PyO3 will provide you with a token. (The GIL is already held when PyO3 calls your function or method, hence we can just pass on the token if you request it.)

**Reply by MohamedAbdeen21** [reply]
I was looking for a way without the token. I totally forgot about the Py token provided for pymethods, I guess this can make things a bit easier.
